File: plugins/plugin-msteams/typescript/src/providers/chatState.ts
import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";

export const CHAT_STATE_PROVIDER = "msteams_chat_state";

/**
 * Provider that exposes MS Teams chat context including conversation ID,
 * user ID, conversation type, and tenant information.
 */
export const chatStateProvider: Provider = {
  name: CHAT_STATE_PROVIDER,
  description:
    "Provides Microsoft Teams chat context including conversation ID, user ID, conversation type, and tenant information",
  dynamic: true,

  get: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state: State,
  ): Promise<ProviderResult> => {
    const conversationId = message.content?.conversationId as
      | string
      | undefined;
    const userId = message.content?.userId as string | undefined;
    const tenantId = message.content?.tenantId as string | undefined;
    const conversationType = message.content?.conversationType as
      | string
      | undefined;
    const activityId = message.content?.activityId as string | undefined;
    const roomId = message.roomId;

    const isPersonal = conversationType === "personal";
    const isGroupChat = conversationType === "groupChat";
    const isChannel = conversationType === "channel";

    const data = {
      conversationId,
      userId,
      tenantId,
      conversationType,
      activityId,
      roomId,
      isPersonal,
      isGroupChat,
      isChannel,
    };

    const values: Record<string, string> = {
      conversation_id: conversationId || "",
      user_id: userId || "",
      tenant_id: tenantId || "",
      conversation_type: conversationType || "",
      activity_id: activityId || "",
      room_id: roomId || "",
      is_personal: isPersonal.toString(),
      is_group_chat: isGroupChat.toString(),
      is_channel: isChannel.toString(),
    };

    let text = "Microsoft Teams Chat State:\n";
    if (conversationId) {
      text += `Conversation ID: ${conversationId}\n`;
    }
    if (conversationType) {
      text += `Conversation Type: ${conversationType}\n`;
    }
    if (userId) {
      text += `User ID: ${userId}\n`;
    }
    if (tenantId) {
      text += `Tenant ID: ${tenantId}\n`;
    }
    if (activityId) {
      text += `Activity ID: ${activityId}\n`;
    }

    return {
      data,
      values,
      text,
    };
  },
};

export const CONVERSATION_MEMBERS_PROVIDER = "msteams_conversation_members";

/**
 * Provider that exposes information about members in the current conversation.
 */
export const conversationMembersProvider: Provider = {
  name: CONVERSATION_MEMBERS_PROVIDER,
  description:
    "Provides information about members in the current Microsoft Teams conversation",
  dynamic: true,

  get: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state: State,
  ): Promise<ProviderResult> => {
    const members = message.content?.members as
      | Array<{
          id: string;
          name?: string;
          email?: string;
        }>
      | undefined;

    const memberCount = members?.length ?? 0;

    const data = {
      members: members ?? [],
      memberCount,
    };

    const values: Record<string, string> = {
      member_count: memberCount.toString(),
      member_names: members?.map((m) => m.name || m.id).join(", ") || "",
    };

    let text = "Microsoft Teams Conversation Members:\n";
    text += `Member Count: ${memberCount}\n`;
    if (members && members.length > 0) {
      text += "Members:\n";
      for (const member of members) {
        text += `- ${member.name || member.id}${member.email ? ` (${member.email})` : ""}\n`;
      }
    }

    return {
      data,
      values,
      text,
    };
  },
};

export const TEAM_INFO_PROVIDER = "msteams_team_info";

/**
 * Provider that exposes information about the current Teams team/channel.
 */
export const teamInfoProvider: Provider = {
  name: TEAM_INFO_PROVIDER,
  description:
    "Provides information about the current Microsoft Teams team and channel",
  dynamic: true,

  get: async (
    _runtime: IAgentRuntime,
    message: Memory,
    _state: State,
  ): Promise<ProviderResult> => {
    const teamId = message.content?.teamId as string | undefined;
    const teamName = message.content?.teamName as string | undefined;
    const channelId = message.content?.channelId as string | undefined;
    const channelName = message.content?.channelName as string | undefined;
    const isThread = Boolean(message.content?.threadId);

    const data = {
      teamId,
      teamName,
      channelId,
      channelName,
      isThread,
    };

    const values: Record<string, string> = {
      team_id: teamId || "",
      team_name: teamName || "",
      channel_id: channelId || "",
      channel_name: channelName || "",
      is_thread: isThread.toString(),
    };

    let text = "Microsoft Teams Team Info:\n";
    if (teamName || teamId) {
      text += `Team: ${teamName || teamId}\n`;
    }
    if (channelName || channelId) {
      text += `Channel: ${channelName || channelId}\n`;
    }
    if (isThread) {
      text += "Message is in a thread\n";
    }

    return {
      data,
      values,
      text,
    };
  },
};

File: plugins/plugin-msteams/rust/src/providers/chat_state.rs
//! Chat state providers for MS Teams.

use async_trait::async_trait;
use serde_json::Value;

use super::{MSTeamsProvider, ProviderContext};

/// Provider that exposes the current conversation/user/tenant context as JSON.
pub struct ChatStateProvider;

#[async_trait]
impl MSTeamsProvider for ChatStateProvider {
    fn name(&self) -> &'static str {
        "msteams_chat_state"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        let is_personal = context
            .conversation_type
            .as_ref()
            .map(|t| t == "personal")
            .unwrap_or(false);
        let is_group_chat = context
            .conversation_type
            .as_ref()
            .map(|t| t == "groupChat")
            .unwrap_or(false);
        let is_channel = context
            .conversation_type
            .as_ref()
            .map(|t| t == "channel")
            .unwrap_or(false);

        serde_json::json!({
            "conversationId": context.conversation_id,
            "userId": context.user_id,
            "tenantId": context.tenant_id,
            "conversationType": context.conversation_type,
            "activityId": context.activity_id,
            "roomId": context.room_id,
            "isPersonal": is_personal,
            "isGroupChat": is_group_chat,
            "isChannel": is_channel
        })
    }
}

/// Provider that exposes information about conversation members.
pub struct ConversationMembersProvider;

#[async_trait]
impl MSTeamsProvider for ConversationMembersProvider {
    fn name(&self) -> &'static str {
        "msteams_conversation_members"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        // In a real implementation, this would fetch members from the service
        serde_json::json!({
            "conversationId": context.conversation_id,
            "members": [],
            "memberCount": 0
        })
    }
}

/// Provider that exposes team and channel information.
pub struct TeamInfoProvider;

#[async_trait]
impl MSTeamsProvider for TeamInfoProvider {
    fn name(&self) -> &'static str {
        "msteams_team_info"
    }

    async fn get(&self, context: &ProviderContext) -> Value {
        let is_channel = context
            .conversation_type
            .as_ref()
            .map(|t| t == "channel")
            .unwrap_or(false);

        serde_json::json!({
            "conversationId": context.conversation_id,
            "tenantId": context.tenant_id,
            "isChannel": is_channel,
            "teamId": null,
            "teamName": null,
            "channelId": if is_channel { context.conversation_id.clone() } else { None },
            "channelName": null
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_chat_state_personal() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            conversation_id: Some("conv-123".to_string()),
            user_id: Some("user-456".to_string()),
            tenant_id: Some("tenant-789".to_string()),
            conversation_type: Some("personal".to_string()),
            activity_id: Some("activity-abc".to_string()),
            room_id: Some("room-xyz".to_string()),
        };

        let state = provider.get(&context).await;
        assert_eq!(state["conversationId"], "conv-123");
        assert_eq!(state["isPersonal"], true);
        assert_eq!(state["isGroupChat"], false);
        assert_eq!(state["isChannel"], false);
    }

    #[tokio::test]
    async fn test_chat_state_group_chat() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            conversation_id: Some("conv-123".to_string()),
            user_id: Some("user-456".to_string()),
            tenant_id: Some("tenant-789".to_string()),
            conversation_type: Some("groupChat".to_string()),
            activity_id: None,
            room_id: None,
        };

        let state = provider.get(&context).await;
        assert_eq!(state["isPersonal"], false);
        assert_eq!(state["isGroupChat"], true);
        assert_eq!(state["isChannel"], false);
    }

    #[tokio::test]
    async fn test_chat_state_channel() {
        let provider = ChatStateProvider;
        let context = ProviderContext {
            conversation_id: Some("19:abc@thread.tacv2".to_string()),
            user_id: Some("user-456".to_string()),
            tenant_id: Some("tenant-789".to_string()),
            conversation_type: Some("channel".to_string()),
            activity_id: None,
            room_id: None,
        };

        let state = provider.get(&context).await;
        assert_eq!(state["isPersonal"], false);
        assert_eq!(state["isGroupChat"], false);
        assert_eq!(state["isChannel"], true);
    }

    #[tokio::test]
    async fn test_team_info_provider() {
        let provider = TeamInfoProvider;
        let context = ProviderContext {
            conversation_id: Some("19:abc@thread.tacv2".to_string()),
            user_id: None,
            tenant_id: Some("tenant-789".to_string()),
            conversation_type: Some("channel".to_string()),
            activity_id: None,
            room_id: None,
        };

        let info = provider.get(&context).await;
        assert_eq!(info["isChannel"], true);
        assert_eq!(info["tenantId"], "tenant-789");
    }
}
